File: 2. AlgoSolution/3. graphTheory/allShortesPath_floydWarshell.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ArrayVisualizerTest
{
    class Program
    {
        static int SIZE = 4;
        static int INF = 99999;

        static int[,] floydWarshell(int[,] G)
        {
            int n = SIZE;
            for (int k = 0; k < n; k++)
                for (int i = 0; i < n; i++)
                    for (int j = 0; j < n; j++)
                        if (G[i, k] + G[k, j] < G[i, j])
                            G[i, j] = G[i, k] + G[k, j];

            return G;
        }

        static void printMatrix(int[,] pIn)
        {
            int n = SIZE;
            for (int i = 0; i < n; i++)
            {
                string text = null;
                for (int j = 0; j < n; j++)
                {
                    if (pIn[i, j] == INF)
                        text += string.Format("{0,6}", "INF");
                    else
                        text += string.Format("{0,6}", pIn[i, j]);
                }
                Console.WriteLine(text);
            }
            Console.WriteLine("");
        }

        static void Main(string[] args)
        {
            /* 
            10
            (0)------->(3)
            |         /|\
            5 |          |
            |          | 1
            \|/         |
            (1)------->(2)
            3           
            */

            int[,] input = new int[4, 4]{ {0,   5,  INF, 10},
	                                     {INF, 0,   3, INF},
	                                     {INF, INF, 0,   1},
	                                     {INF, INF, INF, 0}
	                                   };

            Console.WriteLine("Input Matrix:");
            printMatrix(input);

            int[,] output = floydWarshell(input);
            Console.WriteLine("Output Matrix:");
            printMatrix(output);
        }
    }
}

File: 2. AlgoSolution/3. graphTheory/allShortesPath_floydWarshell.cpp
//http://www.geeksforgeeks.org/dynamic-programming-set-16-floyd-warshall-algorithm/
#include<vector>
#include<algorithm>
#include<iostream>
#include <iomanip>
using namespace std;

#define loop(i,n) for(int i=0;i<(n);i++)
#define MIN(val1, val2) (val1) = min((val1),(val2))

vector<vector<int>> floydWarshell (vector<vector<int>> G)
{
	int X = G.size();
	loop(k,X) loop(i,X) loop(j,X) MIN(G[i][j], G[i][k]+G[k][j]);
	return G;
}


#define V 4
#define INF 99999

void printMatrix(vector<vector<int>>& input)
{
	int X = input.size();
	for (int i = 0; i < X; i++)
	{
		for (int j = 0; j < X; j++)
		{
			if (input[i][j] == INF) cout << setw(6) << "INF";
			else cout << setw(6) << input[i][j];
		}
		cout << endl;
	}
}

int main()
{
	/* 
	10
	(0)------->(3)
	|         /|\
	5 |          |
	|          | 1
	\|/         |
	(1)------->(2)
	3           
	*/
	int graph[V][V] = { {0,   5,  INF, 10},
						{INF, 0,   3, INF},
						{INF, INF, 0,   1},
						{INF, INF, INF, 0}
					  };

	vector<vector<int>> input, output;
	input.resize(V);
	for (int i = 0; i < V; i++)
		for (int j = 0; j < V; j++)
			input[i].push_back(graph[i][j]);

	cout << "Input Matrix:" << endl;
	printMatrix(input);

	output = floydWarshell(input);

	cout << endl << "Output Matrix:" << endl;
	printMatrix(output);

	return 0;
}